java-checks/src/main/java/org/sonar/java/checks/LoopsOnSameSetCheck.java

```diff
@@ -0,0 +1,115 @@
+import org.sonar.plugins.java.api.tree.Tree;
+import org.sonar.squidbridge.annotations.SqaleConstantRemediation;
+import org.sonar.squidbridge.annotations.SqaleSubCharacteristic;
+
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.List;
+import java.util.Map;
+import java.util.Set;
+
+@Rule(
+  key = "S3047",
+  name = "Multiple loops over the same set should be combined",
+  tags = {"performance"},
+  priority = Priority.MAJOR)
+@SqaleSubCharacteristic(RulesDefinition.SubCharacteristics.CPU_EFFICIENCY)
+@SqaleConstantRemediation("20min")
+public class LoopsOnSameSetCheck extends SubscriptionBaseVisitor {
+
+  @Override
+  public List<Tree.Kind> nodesToVisit() {
+    // raise issues only when they appear in the same block
+    return ImmutableList.of(Tree.Kind.BLOCK);
+  }
+
+  @Override
+  public void visitNode(Tree tree) {
+    Map<Symbol, Integer> forEachSymbols = new HashMap<>();
+    Set<Tree> forEachIterables = new HashSet<>();
+    for (Tree item : ((BlockTree) tree).body()) {
+      if (item.is(Tree.Kind.FOR_EACH_STATEMENT)) {
+        checkForEach(forEachSymbols, forEachIterables, (ForEachStatement) item);
+      } else if (item.is(Tree.Kind.EXPRESSION_STATEMENT)) {
+        invalidateReassignedSymbols(((ExpressionStatementTree) item).expression(), forEachSymbols);
+      }
+    }
+  }
+
+  private static void invalidateReassignedSymbols(ExpressionTree given, Map<Symbol, Integer> forEachSymbols) {
+    ExpressionTree expression = ExpressionsHelper.skipParentheses(given);
+    if (expression.is(Tree.Kind.ASSIGNMENT)) {
+      ExpressionTree variable = ExpressionsHelper.skipParentheses(((AssignmentExpressionTree) expression).variable());
+      if (variable.is(Tree.Kind.IDENTIFIER)) {
+        forEachSymbols.remove(((IdentifierTree) variable).symbol());
+      }
+    }
+  }
+
+  private void checkForEach(Map<Symbol, Integer> forEachSymbols, Set<Tree> forEachIterables, ForEachStatement item) {
+    ExpressionTree expressionTree = ExpressionsHelper.skipParentheses(item.expression());
+    if (expressionTree.is(Tree.Kind.IDENTIFIER)) {
+      checkForEachIdentifier(forEachSymbols, (IdentifierTree) expressionTree);
+    } else {
+      checkForEachExpression(forEachIterables, expressionTree);
+    }
+  }
+
+  private void checkForEachExpression(Set<Tree> forEachIterables, ExpressionTree expressionTree) {
+    for (Tree subTree : forEachIterables) {
+      if (SyntacticEquivalence.areEquivalent(expressionTree, subTree)) {
+        addIssue(expressionTree, ((JavaTree) subTree).getLine());
+        return;
+      }
+    }
+    forEachIterables.add(expressionTree);
+  }
+
+  private void checkForEachIdentifier(Map<Symbol, Integer> forEachSymbols, IdentifierTree node) {
+    Symbol symbol = node.symbol();
+    if (forEachSymbols.containsKey(symbol)) {
+      addIssue(node, forEachSymbols.get(symbol));
+    } else {
+      forEachSymbols.put(symbol, ((JavaTree) node).getLine());
```

Comment:
Instead of casting to a `JavaTree` to get the line , preferably use `node.identifierToken().line()` as it will improve the efficiency.